### ISSUE TITLE
Iteratively refine delaboration until roundtrip

### ISSUE THREAD

**Issue by Kha** [leanprover/lean4#368]
We want to produce minimally verbose yet "sound" output from the delaborator that passes re-elaboration. The delaborator already supports annotating subterms with pp options, either inline via `Expr.mdata` or out-of-line with a map indexed by synthetic term positions generated by a first run, but the hard part is generating these annotations e.g. from re-elaboration error messages.

**Comment by Kha** [comment]
Note that roundtripping the parenthesizer and formatter should already be "sound": https://github.com/leanprover/lean4/blob/master/tests/lean/run/Reparen.lean https://github.com/leanprover/lean4/blob/master/tests/lean/Reformat.lean

**Comment by dselsam** [comment]
Noting a few delaborator issues/suggestions here. Most need not be supported explicitly if the delaborator has support for roundtripping-as-necessary. Others may only be useful for porting, and may not need to be built-in to Lean4.

1. It would be nice to have the option `pp.motives` to print (ideally beta-reduced) motives. Conveniently, motives are named `motive`, and so this is easy to support by extending `ParamKind.implicit` with a `Name` and checking for the name/option in `delabAppImplicit`. 
2. It would be nice to have the option `pp.pi_match_motives` to print motives of matches if they are `pi`-types. 
3. Currently, default-argument delaboration is relatively weak. Specifically, only the last default argument can be omitted. This can be strengthened by weaving more information through `delabAppImplicit`, e.g. whether or not a default argument has been skipped yet, and whether there has been a named argument yet.
4. Implicit-lambda delaboration is inconsistent with the elaborator. For example:
```lean
variable {α : Type u} {m : Type u → Type v}
def foo [Monad m] (x : m α) : m α := x
#print foo
```
makes the `[Monad m]` argument implicit, since it does not appear in the downstream syntax:
```lean
def foo.{u, v} : {α : Type u} → {m : Type u → Type v} → [inst : Monad m] → m α → m α :=
fun {α : Type u} {m : Type u → Type v} (x : m α) => x
```
which cannot be elaborated (`x` is expected to have type `m α → m α`).
5. Terms where `lambda`/`pi` binder-infos disagree cannot be re-elaborated. For example:
```lean
theorem foo : ∀ {x : Type}, Type := @λ (x : Type) => x
#print foo
```
yields
```lean
theorem foo : {x : Type} → Type :=
fun (x : Type) => x
```
which cannot be elaborated due to the clashing binder types.
6. The formatter seems somewhat inconsistent with style conventions. For example,
```lean
def foo : Unit → StateM Nat Nat := fun _ => do
  let x ← get
  pure x
```
prints to:
```lean
def foo : Unit → StateM Nat Nat :=
fun (x : Unit) =>
  do 
    let x ← get 
    pure x
```
7. For porting (and perhaps for other use-cases), it would be much more convenient if the `fun`s in a definition body were placed into the header, so that instead of printing 
```lean
def foo.{u, v} : {α : Type u} → {m : Type u → Type v} → [inst : Monad m] → m α → m α :=
fun {α : Type u} {m : Type u → Type v} (x : m α) => x
```
it printed
```lean
def foo.{u, v} {α : Type u} {m : Type u → Type v} [inst : Monad m] (x : m α) : m α :=
  x
```
8. Also for porting, it would be great to have an option `pp.skip_variables` to indicate that arguments and inst-implicit arguments that were already declared as `variable`s with the exact names and types were omitted from definition headers.

**Comment by dselsam** [comment]
9. It may also be nice to have a 
```lean
def idSyntax {α : Sort u} (stx : Syntax) (x : α) : α := x
```
with special delaborator support to print `stx` instead of the regular term. This is easy to write outside of Lean, but it might be nice for some core Lean tactics to make use of this, so e.g. a future `linarith` proof can be pretty-printed as `(by )linarith`.

**Comment by Kha** [comment]
I'm happy to fix any issues in the elaborator, though for the new options it might make more sense to put them in the (temporary) mathport fork as you suggested.



> For porting (and perhaps for other use-cases), it would be much more convenient if the `fun`s in a definition body were placed into the header

This is related to https://github.com/leanprover/lean4/issues/375.

> 9. (`idSyntax`)

I'm not sure reprinting proofs is of any concern for mainline Lean.

**Comment by Kha** [comment]
Effectively solved by #575
